google/cloud/aiplatform/prediction/local_model.py

```diff
@@ -0,0 +1,285 @@
+# -*- coding: utf-8 -*-
+
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from typing import Dict, Optional, Sequence
+
+from google.auth import credentials as auth_credentials
+
+from google.cloud import aiplatform
+from google.cloud.aiplatform import explain
+from google.cloud.aiplatform import models
+
+from google.cloud.aiplatform.compat.types import (
+    model as gca_model_compat,
+    env_var as gca_env_var_compat,
+)
+
+
+class LocalModel:
```

Comment:
Is "local model" the best term here?
Usually, when people talk about a local model, they mean a local model file (XGBoost, Scikit-learn, Tensorflow).
Is this class related to local model files?

Reply:
It aims to be analogous to the local version of Vertex model resource. It will have artifacts while we deploy is to a local endpoint.

Reply:
How would the CUJ for "deploy local model file to local endpoint" scenario look like for the user?

Currently we support a very simple and user-friendly interface for Vertex Models:

```python
model = aiplatform.Model.upload_tensorflow_saved_model("iris.tensorflow_model.SavedModel")
```

Reply:
You could see this PR https://github.com/googleapis/python-aiplatform/pull/1005.

An example workflow should be like.
```
# Builds a custom container.
local_model = LocalModel.create_cpr_model(
    "path/to/user_code",
    "us-docker.pkg.dev/my-project/my-image:0223",
    predictor=CustomPredictor,
    requirements_path=os.path.join(_USER_CODE_DIR, _REQUIREMENTS_FILE),
)
# Locally tests the container.
with local_model.deploy_to_local_endpoint(
    artifact_uri=_ARTIFACT_URI,
    credential_path=os.environ.get("GOOGLE_APPLICATION_CREDENTIALS"),
) as local_endpoint:
    local_predict_response = local_endpoint.predict(
        request=f'{{"instances": [[1, 2, 3, 4]]}}',
        headers={"Content-Type": "application/json"},
    )
# Pushes the built image, uploads, and deploys.
local_model.push_image()
model = local_model.upload("my_custom_model", artifact_uri=_ARTIFACT_URI)
model.deploy(machine_type="n1-standard-2")
```

The main difference is that in this workflow, users are creating a new image and deploy with that built image. But in the sample you mentioned, users are reusing our pre-built images.